text/trace/0192-messaging-semantic-conventions-spec.md

```diff
@@ -0,0 +1,165 @@
+Publish -> | INTERMEDIARY | -> Receive
+           +--------------+
+```
+
+1. The producer creates a message.
+2. The producer publishes the message to an intermediary.
+3. The consumer receives the message from an intermediary.
+4. The consumer processes the message.
+5. The consumer settles the message by notifying the intermediary that the
+   message was processed. In some cases (fire-and-forget), the settlement stage
+   does not exist.
+
+The semantic conventions described below define how to model those stages in
+traces, how to propagate context, and how to enrich traces with attributes.
+
+### Context propagation
+
+Two layers of context propagation need to be distinguished for messaging
+workflows:
+
+1. The _creation context layer_ allows correlating the producer with the
+   consumers of a message, regardless of intermediary instrumentation. The
+   creation context is created by the producer and must be propagated to the
+   consumers. It must not be altered by intermediaries.
+
+   This layer helps to model dependencies between producers and consumers,
+   regardless of the underlying messaging transport mechanism and its
+   instrumentation.
+2. The _transport context layer_ allows correlating the producer and the
+   consumer with an intermediary. It also allows to correlate multiple
+   intermediaries among each other. The transport context can be changed by
+   intermediaries, according to intermediary instrumentations. Intermediaries
+   that are not instrumented might simply drop the transport context.
+
+   This layer helps to gain insights into details of the message transport.
+
+A producer SHOULD attach a creation context to each message. The creation context
+SHOULD be attached in a way so that it is not changed by intermediaries. A
+producer MAY propagate a transport context to an intermediary.  An
+intermediary MAY propagate a transport context to a consumer.
+
+### Trace structure, names, and attributes
+
+#### Consumer
+
+For many use cases, it is not possible to rely on the presence of "Process"
+spans for correlating producer with consumer traces: there are cases where a
+dedicated processing operation cannot be identified, or where processing
+happens in a different trace. Furthermore, processing operations often are not
+covered by messaging libraries and SDKs, but take place in application code.
+Consistently creating spans for "Processing" operations would require either
+effort from the application owner to correctly instrument those operations, or
+additional capabilities of messaging libraries and SDKs (e. g. hooks for
+processing callbacks, which can then be instrumented by the libraries or SDKs).
+
+While it is possible to create "Process" spans and correlate those with
+consumer traces in certain cases, this is not something that can be generally
+required. Therefore, it is more feasible to require the creation of "Deliver"
+spans to correlate producer with consumer traces. A "Deliver" span must link
```

Comment:
linking to the create span is problematic for poll-based messaging systems because links can only be added when span starts.

Is the proposal here to start delivery span when receive is done and messages are pulled from the intermediary? 

If it still traces the whole `client.receive` call, then we should 

- **Option 1: describe how the span is delayed.** Note that in this case any nested transport spans are not children of the Delivery and any nested logs can't be correlated.
  - client.receive starts, instrumentation stores timestamp
  - any logs here are not correlated (have ambient contexT)
  - transport spans (are children of ambient context, likely to be root)  
  - *Delivery* span - starts when pull-based `client.receive` ends (so *Create* contexts are known), instrumentation may 'fix' timestamp to pretend it started earlier

- **Option 2: allow links to be added after span starts**, but it gets us into sampling conversation.
- **Option 3: add more spans** to guarantee mixture of parent/child and links relationships, e.g.:
   
  *Receive* span - describes pull-based `client.receive()` call (no links on it)
     - nested transport spans (children of *Receive*)
     - any logs here are correlated to *Receive*
     - *Delivery* span - describes the short delivery aspect of it,  **links to *Create***

I think option 2 is better in the long-term as it keeps instrumentation to the minimum: it's less verbose (cheaper for users) and has better performance.

Reply:
I've tried to capture pull semantics in [this doc](https://docs.google.com/document/d/1BhYvoHsd-lfiwNtJ6dUIBiP4fxyBmI8FLj1C6xR-kFg/edit?usp=sharing) including these options in more detail. Let's discuss it over the next meeting once we have a quorum.

Reply:
Thanks for coming up with that detailed analysis, let's discuss this in the next workgroup meeting.

Reply:
I currently wrote conventions in the OTEP for going with Option 2. I also created a PR for allowing adding links after span creation: https://github.com/open-telemetry/opentelemetry-specification/pull/2278

If that PR is not approved, we will re-iterate and discuss alternative solutions.